<a href="https://colab.research.google.com/github/robitussin/CCRNFLRL_EXERCISES/blob/main/Exercise2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2

In [537]:
from math import sqrt, log
import pandas as pd

## Step 1: Enter your data

Add a new row after each pull: Pull number, Machine played ("A", "B", "C"), and Reward earned.

In [538]:
data = pd.DataFrame({
    "Pull": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
             16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28,
             29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50],         # Pull number
    "Machine": ["A", "B", "C", "A", "B", "C", "A", "B", "C", "A", "A", "A", "A",
                "A", "B", "C", "A", "A", "B", "C", "A", "A", "A", "A", "B", "C",
                "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A",
                "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A"], # Which machine was played
    "Reward": [0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 0,
               0, 0, 0, 2, 0, 0, 0, 0, 0, 5, 0, 5, 0, 0, 0, 10, 0,
               0, 0, 0, 0, 80, 0, 0, 0, 10, 215, 0, 2, 0, 0, 0, 5]         # Reward received from that machine
})

# Step 2: Calculate UCB

In [539]:
def calculate_ucb(df):
    results = []
    total_pulls = len(df)
    machines = sorted(df["Machine"].unique())

    for machine in machines:
        machine_data = df[df["Machine"] == machine]
        n_j = len(machine_data)
        avg_reward = machine_data["Reward"].mean()
        exploration_bonus = sqrt((2 * log(total_pulls)) / n_j) if n_j > 0 else float("inf")
        ucb_score = avg_reward + exploration_bonus
        results.append({
            "Machine": machine,
            "Pulls": n_j,
            "Avg Reward": round(avg_reward, 2),
            "Exploration Bonus": round(exploration_bonus, 4),
            "UCB Score": round(ucb_score, 4)
        })
    return pd.DataFrame(results).sort_values(by="UCB Score", ascending=False)


# Step 3: Calculate and display


In [540]:
ucb_table = calculate_ucb(data)
print("📊 UCB Results Table")
print(ucb_table)

📊 UCB Results Table
  Machine  Pulls  Avg Reward  Exploration Bonus  UCB Score
0       A     38        8.89             0.4538     9.3485
1       B      6        0.00             1.1419     1.1419
2       C      6        0.00             1.1419     1.1419


# Step 4: Next machine suggestion

In [541]:
best_machine = ucb_table.iloc[0]["Machine"]
print(f"\n🎯 Recommended next machine to play: {best_machine}")


🎯 Recommended next machine to play: A


# After 50 Pulls Machine A (Iguana) is the most profitable

# Machine A getting an Average Reward of 8.89 while Machines B and C got 0.00